src/zkevm_specs/evm_circuit/execution/begin_tx.py

```diff
@@ -91,16 +104,94 @@ def begin_tx(instruction: Instruction):
         MAX_N_BYTES,
     )
     invalid_tx = 1 - (1 - balance_not_enough) * (1 - gas_not_enough) * (is_nonce_valid)
+
     # prover should not give incorrect is_tx_invalid flag.
     instruction.constrain_equal(is_tx_invalid, invalid_tx)
 
     if tx_is_create == 1:
-        # TODO: Verify created address
-        # code_hash represents the contract creation code
-        # 1. In the case of root call, code_hash is the hash of the tx calldata.
-        # 2. In the case of internal call, code_hash is the hash of the chunk of
-        #    caller memory corresponding to the contract init code.
-        raise NotImplementedError
+        if is_tx_invalid == FQ(1) or tx_call_data_length == 0:
+            # Make sure tx is persistent
+            instruction.constrain_equal(reversion_info.is_persistent, FQ(1))
+
+            # Do step state transition
+            instruction.constrain_equal(instruction.next.execution_state, ExecutionState.EndTx)
+            instruction.constrain_step_state_transition(
+                rw_counter=Transition.delta(9), call_id=Transition.to(call_id)
+            )
+        else:
+            # Expected behabeur
+            # - If initcode does not RETRUN, contract is created empty and value transferred
+            # - If initcode is invalid bytecode or reverts, contract is not created and value not transferred
+
+            # Get code hash of tx calldata
+
+            copy_rwc_inc, rlc_acc = instruction.copy_lookup(
+                tx_id,  # src_id
+                CopyDataTypeTag.TxCalldata,  # src_type
+                call_id,  # dst_id
+                CopyDataTypeTag.RlcAcc,  # dst_type
+                FQ.zero(),  # src_addr
+                tx_call_data_length,  # src_addr_boundary
+                FQ.zero(),  # dst_addr
+                tx_call_data_length,  # length
+                instruction.curr.rw_counter + instruction.rw_counter_offset,
+            )
+
+            assert copy_rwc_inc == FQ.zero()
+            # no memory involved, no rw counter incremented
+
+            code_hash = instruction.keccak_lookup(tx_call_data_length, rlc_acc)
+            is_empty_code_hash = instruction.is_equal_word(code_hash, Word(EMPTY_CODE_HASH))
```

Comment:
You could also reuse the `tx_call_data_length == 0` check for this, to save some constraints and cells.

Reply:
I guess this constraint is not necessary because `tx_call_data_length` must be a non-zero value in this `else` condition. Since `calldata` is non-zero data, the `code_hash` would not be equal to `EMPTY_CODE_HASH`.